### DISCUSSION TITLE
Edit Cell of Table with Drop Down Selection

### DISCUSSION THREAD

**Discussion by Slanderkin** [zauberzeug/nicegui#675] [Q&A] [answered]
Is it currently possible to utilize either and AGGrid or a Quasar Table in Nicegui to edit a cell by choosing a value from a dropdown table. I know that for the AGgrid you could possibly use:
```Python
{'headerName': 'category', 
'field': 'category', 
'sortable': True, 
'cellEditor': 'agRichSelectCellEditor', 
'cellEditorPopup': True, 
'cellEditorParams': {'values': ["foo", 'bar']}}
```
although I wasn't able to get something like this to work.

I believe with a Quasar Table you could use a [Q Select component ](https://quasar.dev/vue-components/select#Filtering-and-autocomplete) but I'm not quite sure how this would be implemented let alone how you would then grab data from the selects within the table. Any thoughts/direction is much appreciated!

**Comment by falkoschindler** [comment]
That's an interesting question, @Slanderkin! So far I haven't found a solution. But I started to play around with editable QTable cells and would like to share my code.

Using scoped body slots I managed to get the [popup editing example](https://quasar.dev/vue-components/table#popup-editing) to work with NiceGUI. On save a custom "save" event is sent to Python, where we can receive it with the generic event registration `.on()`:

```py
columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'age', 'label': 'Age', 'field': 'age'},
]
rows = [
    {'name': 'Alice', 'age': 18},
    {'name': 'Bob', 'age': 21},
    {'name': 'Carol'},
]
with ui.table(columns=columns, rows=rows, row_key='name').classes('w-72') as table:
    table.add_slot('body', r'''
        <q-tr :props="props">
            <q-td key="name" :props="props">
                {{ props.row.name }}
                <q-popup-edit v-model="props.row.name" v-slot="scope" @save="(v) => $parent.$emit('save', {name: props.row.name, new_name: v})">
                    <q-input v-model="scope.value" dense autofocus counter @keyup.enter="scope.set" />
                </q-popup-edit>
            </q-td>
            <q-td key="age" :props="props">
                {{ props.row.age }}
            </q-td>
        </q-tr>
    ''')
    table.on('save', lambda msg: ui.notify(f'"{msg["args"]["name"]}" has been renamed to "{msg["args"]["new_name"]}"'))
```

Maybe someone can take this a step further and integrate QSelect dropdowns. Once we have examples like this working, we can think about integrating it into the public API.

**Comment by Slanderkin** [comment]
@falkoschindler  So I spent some time working on this and I think I have a basic solution.

```Python
from nicegui import ui

columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'age', 'label': 'Age', 'field': 'age'},
]
rows = [
    {'name': 'Alice', 'age': 18},
    {'name': 'Bob', 'age': 21},
    {'name': 'Carol'},
]

#name_options = ['Alice','Bob','Carol']

with ui.table(columns=columns, rows=rows, row_key='name').classes('w-72') as table:
    table.add_slot('body', r'''
        <q-tr :props="props">
            <q-td key="name" :props="props">
                <q-select v-model="props.row.name" :options=['Alice','Bob','Carol'] label="Standard" @update:model-value="(v) => $parent.$emit('value-change', {value: v})"/>
            </q-td>
            <q-td key="age" :props="props">
                {{ props.row.age }}
            </q-td>
        </q-tr>
    ''')
    table.on('value-change', lambda msg: ui.notify(f'"{msg["args"]["value"]}" has been selected"'))
ui.run(title='QSelect Test', dark=True, native=True)
```

I don't have much experience with raw strings in python so I didn't know of an easy/elegant way to get the name_options list to work instead of hardcoding it in the string but regardless I believe this is the simplest form of a drop down select in a table.

Let me know if you think this is correct or ways to improve it!

**Reply by falkoschindler** [reply]
Great! 😀
Yes, combining raw strings with string interpolation can be tricky. I'd simply concatenate the strings in this case. But there are plenty ways to do it.

```py
table.add_slot('body', r'''
    ...
    <q-select v-model="props.row.name" :options="''' + str(name_options) + r'''" label="Standard" @update:model-value="(v) => $parent.$emit('value-change', {value: v})"/>
    ...
''')
```

**Reply by Slanderkin** [reply]
@falkoschindler I had marked this as the answer, but upon further inspection this doesn't actually change the table's value. The table.rows does not update with the user selection. I tried adding [emit-value](https://quasar.dev/vue-components/select#affecting-model) like mentioned in the q-select docs but that didn't seem to work.
```Python
<q-select v-model="props.row.name" :options=['Alice','Bob','Carol'] label="Standard" emit-value @update:model-value="(v) => $parent.$emit('value-change', {value: v})"/>
```
This can be seen by changing the event capture line to:
```Python
table.on('value-change', lambda msg: print(table.rows))
```

Do you have any thoughts on why the user changes isn't actually being populated?

**Reply by falkoschindler** [reply]
Oh, that's right. We're only consuming events from the client. In order to keep the server side in sync, you'd need to update the table in Python.

**Comment by Slanderkin** [comment, marked as answer]
For the future use of anyone else who wants to do this, here is how I solved this including updating the server:
```Python
from nicegui import ui

columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'age', 'label': 'Age', 'field': 'age'},
]
rows = [
    {'id': 0, 'name': 'Alice', 'age': 18},
    {'id': 1,'name': 'Bob', 'age': 21},
    {'id': 2,'name': 'Carol'},
    {'id': 3,'name': 'Jill'},
]

def edit_table(msg):
    for row in rows:
        if row['id'] == msg['args']['id']:
            row['name'] = msg['args']['name']
    ui.notify(f'Table.rows is now: {table.rows}')

name_options = ['Alice','Bob','Carol']

with ui.table(columns=columns, rows=rows, row_key='name').classes('w-72') as table:
    table.add_slot('body', r'''
        <q-tr :props="props">
            <q-td key="name" :props="props">
                <q-select v-model="props.row.name" :options="''' + str(name_options) + r'''" label="Standard" emit-value @update:model-value="(v) => $parent.$emit('value-change', {id: props.row.id, name: props.row.name, value: v})"/>
            </q-td>
            <q-td key="age" :props="props">
                {{ props.row.age }}
            </q-td>
        </q-tr>
    ''')
    table.on('value-change', lambda msg: edit_table(msg))
ui.run(title='QSelect Test', dark=True, native=True)
```

You can see that editing the local rows variable still makes changes in the table.rows.

**Reply by marcin-koziol** [reply]
How to make different options per row?

**Reply by rodja** [reply]
@marcin-koziol please start another question and link to this one if it's related. Like on Stack Overflow it's much better to have everything separated.
